[AbnormalSecurity] Expand integration commands

wolyslager opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
This PR expands the Abnormal Security Integration to include new commands through which users can interact with new endpoints in the Abnormal API. These commands expose information related to Vendors, Vendor Cases, and Abuse Mailbox submissions that were previously not available to the user through the cortex tool. Additionally, this PR expands the number of fields returned by the threat-details endpoint to provide users with more information. 

## Must have
- [ ] Tests
- [ ] Documentation

wolyslager commented:
Hey @MichaelYochpaz ! I have another PR where I will be implementing the `fetch_incidents()` logic - was curious to know what the preferred approach is. Should I stack the PR on this one or wait until this one is merged before submitting the other one? Let me know :) thanks!

wolyslager commented:
The validation seems to be working on my end locally but not on the circleci run - going to try running the suggested commands 
<img width="329" alt="image" src="https://github.com/demisto/content/assets/29632564/fa7219ad-0188-48ca-97e8-1f705e4d9c2c">

MichaelYochpaz commented:
> Hey @MichaelYochpaz ! I have another PR where I will be implementing the fetch_incidents() logic - was curious to know what the preferred approach is. Should I stack the PR on this one or wait until this one is merged before submitting the other one? Let me know :) thanks!

@wolyslager Hey. If adding the fetch incidents logic isn't urgent, I think it'll be better to do that in a separate PR after merging this one because there will probably be conflicts you'll have to resolve in the other PR once one of them is merged, if you work on both simultaneously.

As for the failing validation - it fails in a step that runs `demisto-sdk secrets --post-commit --ignore-entropy`.

This can be fixed by adding the detected IP & mail (assuming they're fake and aren't actual secrets) to the `Packs/AbnormalSecurity/.secrets-ignore` file.

wolyslager commented:
> > Hey @MichaelYochpaz ! I have another PR where I will be implementing the fetch_incidents() logic - was curious to know what the preferred approach is. Should I stack the PR on this one or wait until this one is merged before submitting the other one? Let me know :) thanks!
> 
> @wolyslager Hey. If adding the fetch incidents logic isn't urgent, I think it'll be better to do that in a separate PR after merging this one because there will probably be conflicts you'll have to resolve in the other PR once one of them is merged, if you work on both simultaneously.
> 
> As for the failing validation - it fails in a step that runs `demisto-sdk secrets --post-commit --ignore-entropy`.
> 
> This can be fixed by adding the detected IP & mail (assuming they're fake and aren't actual secrets) to the `Packs/AbnormalSecurity/.secrets-ignore` file.

Awesome - good to know. So our our target to have both this PR and the fetch_incidents PR merged is before end of September. If you think that is doable doing the PRs sequentially then that sounds good to me! If you would prefer to have all of the changes in one PR and you think that would be faster happy to do that as well.

MichaelYochpaz commented:
> Awesome - good to know. So our our target to have both this PR and the fetch_incidents PR merged is before end of September. If you think that is doable doing the PRs sequentially then that sounds good to me! If you would prefer to have all of the changes in one PR and you think that would be faster happy to do that as well.

 I think adding all the changes to this PR (adding the fetch implementation here to this PR) will be the fastest then.
Make sure to update the current release notes with the changes you're merging in

wolyslager commented:
> > Awesome - good to know. So our our target to have both this PR and the fetch_incidents PR merged is before end of September. If you think that is doable doing the PRs sequentially then that sounds good to me! If you would prefer to have all of the changes in one PR and you think that would be faster happy to do that as well.
> 
> I think adding all the changes to this PR (adding the fetch implementation here to this PR) will be the fastest then. Make sure to update the current release notes with the changes you're merging in

@MichaelYochpaz perfect, just added the remaining changes to this PR - I really appreciate it!

MichaelYochpaz commented:
@wolyslager I see that you've pushed the fetch-incidents changes. Great.

A small update - PR review will be slightly delayed since it is the holiday season in Israel, and we have a holiday this weekend (15th-17th of September), I apologize in advance.

It seems like [some validations are failing](https://app.circleci.com/pipelines/github/demisto/content/302778/workflows/ae587252-b0f8-4c2a-8318-e7dc8cc118ea/jobs/719713), and that [the unit-tests coverage isn't high enough](https://app.circleci.com/pipelines/github/demisto/content/302778/workflows/ae587252-b0f8-4c2a-8318-e7dc8cc118ea/jobs/719709), so you can work on those in the meantime.

wolyslager commented:
> @wolyslager I see that you've pushed the fetch-incidents changes. Great.
> 
> A small update - PR review will be slightly delayed since it is the holiday season in Israel, and we have a holiday this weekend (15th-17th of September), I apologize in advance.
> 
> It seems like [some validations are failing](https://app.circleci.com/pipelines/github/demisto/content/302778/workflows/ae587252-b0f8-4c2a-8318-e7dc8cc118ea/jobs/719713), and that [the unit-tests coverage isn't high enough](https://app.circleci.com/pipelines/github/demisto/content/302778/workflows/ae587252-b0f8-4c2a-8318-e7dc8cc118ea/jobs/719709), so you can work on those in the meantime.

Sounds good @MichaelYochpaz - enjoy the Holiday! I will work on these changes in the meantime :)

MichaelYochpaz commented on Packs/AbnormalSecurity/Integrations/AbnormalSecurity/AbnormalSecurity.py:
```diff
@@ -529,6 +620,207 @@ def submit_false_positive_report_command(client, args):
 
 
+def get_a_list_of_vendors_command(client, args):
+    page_size = str(args.get('page_size', ''))
+    page_number = str(args.get('page_number', ''))
```
For int-based parameters, we have a `arg_to_number` CommonServerPython function that prints proper errors for users if an invalid input is used (anything that isn't a number). It returns an int, though.

I'd recommend passing int parameters as an int type between functions, not a string. Then convert them when sending HTTP requests (I think it might be done automatically by `requests` though, so you might not even need to do that).

The same note applies for other cases in other functions below.

MichaelYochpaz reviewed (commented):
Hey, looks pretty good overall, good job :)
Left a few minor notes.

wolyslager commented:
> Hey, looks pretty good overall, good job :) Left a few minor notes.

Thanks @MichaelYochpaz ! Addressed your comments - I appreciate the thoughtful review. Let me know if something is still not up to par and I'll take care of it. Look forward to getting this merged in

MichaelYochpaz commented:
@wolyslager Good job :)

Please run `demisto-sdk generate-docs -i "Packs/AbnormalSecurity/Integrations/AbnormalSecurity/AbnormalSecurity.yml"` to update the README following your changes (go over the changes the SDK did. Sometimes it requires some manual tweaking / changes).

[The unit-tests pipeline also seems to fail in a few steps](https://app.circleci.com/pipelines/github/demisto/content/304493/workflows/efa6154c-dc0a-438c-bb35-e4e3506a92dc/jobs/720577).

The "Infrastructure testing" step there is an issue we had that's unrelated to your PR. 
Merge from master and it should be fixed.

The "Run Unit Testing And Lint - Docker Image:from-yml" has a few unit-tests for this integration failing that you'll have to address / fix.

There's also the "Generate coverage reports" step failing there, since there are now additional functions but not tests, so the coverage percentage dropped.
You can, however, add `# pragma: nocover` for the `main` function, since it doesn't require a unit-test. It should increase your coverage since those lines won't be counted. Maybe that will get you over the required amount.

Should look like:

```python
def main():  # pragma: no cover
```

Feel free to reach out to me on Slack if you need any help / have questions.

wolyslager commented:
> @wolyslager Good job :)
> 
> Please run `demisto-sdk generate-docs -i "Packs/AbnormalSecurity/Integrations/AbnormalSecurity/AbnormalSecurity.yml"` to update the README following your changes (go over the changes the SDK did. Sometimes it requires some manual tweaking / changes).
> 
> [The unit-tests pipeline also seems to fail in a few steps](https://app.circleci.com/pipelines/github/demisto/content/304493/workflows/efa6154c-dc0a-438c-bb35-e4e3506a92dc/jobs/720577).
> 
> The "Infrastructure testing" step there is an issue we had that's unrelated to your PR. Merge from master and it should be fixed.
> 
> The "Run Unit Testing And Lint - Docker Image:from-yml" has a few unit-tests for this integration failing that you'll have to address / fix.
> 
> There's also the "Generate coverage reports" step failing there, since there are now additional functions but not tests, so the coverage percentage dropped. You can, however, add `# pragma: nocover` for the `main` function, since it doesn't require a unit-test. It should increase your coverage since those lines won't be counted. Maybe that will get you over the required amount.
> 
> Should look like:
> 
> ```python
> def main():  # pragma: no cover
> ```
> 
> Feel free to reach out to me on Slack if you need any help / have questions.

@MichaelYochpaz I do have a few questions, how can I reach out through slack?

MichaelYochpaz commented:
@wolyslager You can join our DFIR slack here: https://start.paloaltonetworks.com/join-our-slack-community.
Once joined, just look up my name up and message me.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=29558) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=29558) before we can accept your contribution.<br/>**13** out of **15** committers have signed the CLA.<br/><br/>:white_check_mark: samuelFain<br/>:white_check_mark: wolyslager<br/>:white_check_mark: Shellyber<br/>:white_check_mark: eyalpalo<br/>:white_check_mark: BEAdi<br/>:white_check_mark: yasta5<br/>:white_check_mark: eepstain<br/>:white_check_mark: ShirleyDenkberg<br/>:white_check_mark: ilaner<br/>:white_check_mark: content-bot<br/>:white_check_mark: GuyAfik<br/>:white_check_mark: darkushin<br/>:white_check_mark: jlevypaloalto<br/>:x: tkatzir<br/>:x: kobymeir<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=29558) it.</sub>

wolyslager commented:
@MichaelYochpaz Unsure about how to resolve the validations issue there, doesn't seem directly related to the application code - but let me know if its on my end and I will fix it.

MichaelYochpaz commented:
Hey @wolyslager, note [the message here](https://github.com/demisto/content/pull/29558/files#diff-021da102fcff1a613c1c00df7f874063f9232683b066b42b26c056a95fc9af71), you'll need to update the Docker image version mentioned in the release-notes (you've updated it on the YAML file, but not on the release-notes).

Other than that PR looks great, good job :)
I'll email you about scheduling a demo.

MichaelYochpaz reviewed (commented):
Hey, demo went well, good job :)

A few notes following the demo:
1. Let's remove pagination info from the returned context data, since it's pretty useless for users. The best way, in my opinion, would be to override and wrap `_http_request` and just remove the pagination key from responses ([see an example here](https://github.com/demisto/content/blob/master/Packs/Rapid7_Nexpose/Integrations/Rapid7_Nexpose/Rapid7_Nexpose.py#L149-L156)), but it's up to you to decide how you'd want to implement it.
2. Let's add a warning on the problematic command (I believe it was `abnormal-security-get-threat`?), both on its description (YAML file) and README (you can just regenerate it from the YAML file after updating it) saying that the endpoint in use for this command is currently having issues, and the API call might fail.
3. Left a note about the `fetch incidents` issue we discussed. Will check about that and give you an update.

wolyslager commented on Packs/AbnormalSecurity/Integrations/AbnormalSecurity/AbnormalSecurity.py:
```diff
@@ -172,6 +197,65 @@ def submit_false_positive_report_request(self, portal_link):
+
+        return response
+
+    def _remove_keys_from_response(self, response, keys_to_remove):
```
For now I think this should suffice - did not want to add more complexity w/out necessity (so no need for a recursive pop here)

wolyslager commented:
@MichaelYochpaz As we discussed, the following changes have been made:
- Added checkboxes in the Integrations Settings for users to specify which endpoints they want to enable for fetch_incidents
- The get_threats endpoint has been stabilized and is now performant (we don't anticipate more issues with this) 
- The fetch_incidents command will fail if any individual request fails, which removes confusion from the users as you mentioned. 
- We remove the pageNumber keys from the endpoints that contain it. 

Let me know if something is out of place, hoping to get this merged as soon as we can 😄  - thank you!

MichaelYochpaz commented:
@wolyslager Please merge from master and push to this branch. This should fix the current PR issue (it's mostly due to an issue on our side that happens from time to time, not because you did something wrong).

wolyslager commented:
> @wolyslager Please merge from master and push to this branch. This should fix the current PR issue (it's mostly due to an issue on our side that happens from time to time, not because you did something wrong).

@MichaelYochpaz I merged from master and pushed - it seems to be adding in other changes that I did not make for some reason - any idea how to fix this?

MichaelYochpaz commented:
Closing, as this PR was reopened on https://github.com/demisto/content/pull/29994
